### PR TITLE
[RFR] Add prefixed version

### PR DESCRIPTION
**Pull request by eduardoboucas**
This is attempt to alleviate the problems mentioned in #32. What do you think @HugoGiraudel?

### CODE REVIEW

**Comment by scottkellum** [comment]
I have had specific mixins as the core mixin in the past and more generic mixins as an alias. Core functions and mixins should always be namespaced.

ex:

```scss
@mixin include-media($media) {
  @content
}

@mixin media($media) {
  @include include-media($media) {
    @content
  }
}
```

This way both a rigidly namespaced mixin is part of core and easily maintainable and an abbreviated generic mixin aliases to the more bulletproof mixin.

File: dist-prefixed/_include-media.scss [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+
+
+///
+/// Log a message either with `@error` if supported 
+/// else with `@warn`, using `feature-exists('at-error')`
+/// to detect support.
+/// 
+/// @param {String} $message - Message to log
+///
+@function log($message) {
```
Comment (scottkellum) [inline dist-prefixed/_include-media.scss]: This is at risk of running into namespace errors. Especially with math functions for `log()`

File: dist-prefixed/_include-media.scss [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+/// @example scss - Extend both configuration maps
+///  @include im-media-context(('custom': 678px), ('all': 'all')) {
+///    .foo {
+///      @include im-media('all', '>phone', '<=custom') {
+///       // ...
+///      }
+///    }
+///  }
+///
+@mixin im-media-context($tweakpoints: (), $tweak-media-expressions: ()) {
```
Comment (scottkellum) [inline dist-prefixed/_include-media.scss]: :+1: well scoped. Would be nice to have this prefix on all non-user facing utility functions and mixins

Sass is also case sensitive so @snugug has experimented with uppercase to scope all non-user facing prefixes.

File: dist-prefixed/_include-media.scss (hunk 1 above)
Comment (HugoGiraudel) [inline dist-prefixed/_include-media.scss]: True.

File: dist-prefixed/_include-media.scss (hunk 2 above)
Comment (eduardoboucas) [inline dist-prefixed/_include-media.scss]: Yep, that's part of the plan.

**Comment by kaelig** [comment]
Seems very much like `over-[engineering|kill]`, but I might be missing something.

I admire the effort, though.

**Comment by gilbox** [comment]
@kaelig to provide you with some perspective, the entire css ecosystem of scalable solution seems over-engineered because css and sass are not good solutions to build scalable solutions. It's why many people are turning to CSS in JS-based solutions but there are challenges getting that to work with RWD and of course many existing projects are better off putting :lipstick: on their old :pig:

**Comment by kaelig** [comment]
Thanks @gilbox.

Instead of relying on a build step, wouldn't it be simpler to prefix manually all variables, functions and mixins and then publish a new major version?

**Comment by HugoGiraudel** [comment]
Please do not merge this. I want to review it carefully.

**Comment by scottkellum** [comment]
@kaelig :+1: don’t even need a major version bump as generic namespace can be preserved with aliases as stated in https://github.com/eduardoboucas/include-media/pull/94#issuecomment-126375365

**Comment by kaelig** [comment]
With that step:
- you have to double the integration tests (one for each version) if you really want to be thorough
- potentially: maintain that build step
- the project documentation/site/readme gets bigger (showing both versions?)
- Maybe SassDoc gets a lot bigger too?

Why not be opinionated and just release one prefixed version?

**Comment by eduardoboucas** [comment]
@scottkellum:
> This way both a rigidly namespaced mixin is part of core and easily maintainable and an abbreviated generic mixin aliases to the more bulletproof mixin.

@HugoGiraudel suggested that [here](https://github.com/eduardoboucas/include-media/issues/32#issuecomment-123641923), but we ended up deciding against it because that doesn't completely eliminate the problem. If you load include-media after Bourbon, it will still overwrite Bourbon's `media` mixin, so the import order would still be very important and dictate whether both libraries can work in conjunction.

I'm not entirely against this approach and I'm more than happy to go down this route if people think it's best, I just don't want users to think that this solves the problem completely to the point that they don't need to worry about import order because everything will just work fine.

@kaelig:
> Instead of relying on a build step, wouldn't it be simpler to prefix manually all variables, functions and mixins and then publish a new major version?

I understand that's the most bulletproof solution, but as I mentioned in #32 I don't want to break compatibility with everything we've done so far, and more importantly, lose the principle of having a simple and natural syntax across the library by introducing a prefix. Especially when we know that, as @chriseppstein mentioned, this problem will be addressed by the language.

I agree this additional dist file is not a pretty solution, and I think you're right with all the disadvantages you pointed out. My idea behind it was to create a temporary solution until the underlying problems that cause this issue are dealt with upstream.

Interesting discussion though!

**Comment by chriseppstein** [comment]
If you did take @kaelig's solution, it would be possible to build a single import like `@import "include-media/unprefixed";` that would remove the prefixes from all the mixins and functions, etc by simply wrapping them with an unprefixed version of the mixin.

```scss
@mixin foo($args...) {
  @include prefix-foo($args...);
}
```

**Comment by HugoGiraudel** [comment]
Alright, I am just coming back to this. Sorry, I did not have much time to take care of the project lately. Damn, it kills me that we're still talking about this issue... For god sake, it's just about namespacing.

Here is my gut feeling about it: I would like to namespace everything with `im-`. I feel like namespacing is always a good idea, especially with something as simple and short as `im-`. For people not liking the namespace, they still can do:

```scss
@mixin media($args...) {
  @include im-media($args...) {
    @content;
  }
}
```

This way, we prevent any conflict with other libraries or even future Sass functions, and still make it very easy for anybody to get rid of the namespace if they want to. Win-win. Especially since most people don't give a fuck about a 2 letters namespace as far as I can tell. It's not a big deal.

If you really don't want to prefix it @eduardoboucas, I suggest we still prefix everything but the `media` mixin (and possibly the `media-context` mixin). We avoid any conflict with everything but the `media` mixin, and keep the API short and the way you like it. At this point, it means that we don't want to deal with potential conflicts with other libraries and users have to fix it themselves. It can be something as simple as:

```scss
// Import include-media
@import 'include-media';

// Prefix the API (possibly done in another file)
@mixin im-media($args...) {
  @include media($args...) {
     @content:
  }
}

// Import Bourbon
@import 'bourbon';
```

In any case, I don't want to deal with this as a build process step. It sounds wrong to me. This problem is a flaw in Sass language (in some way), and should be dealt with both sensible code and good documentation. Not with a damn Gulp task.

I let you take the decision Eduardo, your move. You have my 2 cents, at least. :)

**Comment by eduardoboucas** [comment]
> Damn, it kills me that we're still talking about this issue... For god sake, it's just about namespacing.

I would agree with this on most cases, and I can see how it may seem irrelevant, but I think it has some importance when the syntax is all we have to offer. We didn't invent the wheel when it comes to breakpoint managers, but what I think we did get right is how simple and natural the syntax is. I was trying to avoid breaking that relationship of familiarity people already have with it, that makes it obvious enough to immediately go `include-media to target a tablet <=> @include media('>tablet')`.

> In any case, everything that is not part of the public API will be prefixed, for sure.

Completely agree. I think all private functions should be namespaced, independently of what we end up deciding.

All in all, I'm happy to go ahead and prefix the mixin as well if we think it's the best way to go about it.

**Comment by eduardoboucas** [comment]
> In any case, I don't want to deal with this as a build process step.

Cool, let's kill that idea then.

**Comment by HugoGiraudel** [comment]
Both @kaelig and @chriseppstein told me I seemed pissed. My apologies for sounding harsh. No harm intended. :)

**Comment by HugoGiraudel** [comment]
Also, @chriseppstein just told me that the following snippet won't work:

```scss
// Import include-media
@import 'include-media';

// Prefix the API (possibly done in another file)
@mixin im-media($args...) {
  @include media($args...) {
     @content;
  }
}

// Import Bourbon
@import 'bourbon';
```

The `media` mixin is resolved when `im-media` is called, not when it’s defined. So one more reason to prefix everything.

**Comment by eduardoboucas** [comment]
This approach is silly, I'm closing this. We'll come back to namespacing on a later episode.

**Pull request closed without merging**